File: python-selenium-pytest-bdd/src/saucedemo_uitests/steps/cart_steps.py
from pytest_bdd import parsers, then, when
from selenium.webdriver.remote.webdriver import WebDriver

from saucedemo_uitests.pages.cart_page import CartPage
from saucedemo_uitests.pages.inventory_page import InventoryPage
from saucedemo_uitests.support.polling import assert_eventually


@then(parsers.parse("the cart badge should show {expected_count:d}"))
def the_cart_badge_should_show(driver: WebDriver, expected_count: int) -> None:
    assert_eventually(lambda: InventoryPage(driver).get_cart_count(), expected_count, "cart badge count")


@when(parsers.parse('I remove "{product_name}" from the cart'))
def remove_product_from_the_cart(driver: WebDriver, product_name: str) -> None:
    InventoryPage(driver).remove_from_cart(product_name)


@when("I start checkout")
def start_checkout(driver: WebDriver) -> None:
    CartPage(driver).start_checkout()


@then("the cart should list the following products:")
def the_cart_should_list_the_following_products(driver: WebDriver, datatable: list[list[str]]) -> None:
    header, *rows = datatable
    name_index = header.index("ProductName")
    expected = sorted(row[name_index] for row in rows)

    assert_eventually(lambda: sorted(CartPage(driver).list_item_names()), expected, "cart item names")

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/steps/CartSteps.java
package com.saucedemo.uitests.steps;

import com.saucedemo.uitests.hooks.ScenarioContext;
import com.saucedemo.uitests.pages.CartPage;
import com.saucedemo.uitests.pages.InventoryPage;
import com.saucedemo.uitests.support.Polling;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import java.util.List;

public class CartSteps {

    private final ScenarioContext context;

    public CartSteps(ScenarioContext context) {
        this.context = context;
    }

    @Then("the cart badge should show {int}")
    public void theCartBadgeShouldShow(int expectedCount) {
        // Polled, not read once: the badge re-renders asynchronously after a click.
        Polling.assertEventually(
                () -> new InventoryPage(context.driver()).cartCount(), expectedCount, "cart badge count");
    }

    @When("I remove {string} from the cart")
    public void iRemoveFromTheCart(String productName) {
        new InventoryPage(context.driver()).removeFromCart(productName);
    }

    @When("I start checkout")
    public void iStartCheckout() {
        new CartPage(context.driver()).startCheckout();
    }

    @Then("the cart should list the following products:")
    public void theCartShouldListTheFollowingProducts(DataTable table) {
        List<String> expected = table.asMaps().stream()
                .map(row -> row.get("ProductName"))
                .sorted()
                .toList();

        Polling.assertEventually(
                () -> new CartPage(context.driver()).listItemNames().stream().sorted().toList(),
                expected,
                "cart item names");
    }
}
